### PR TITLE
Phase2-hgx314 A step toward the V17 version of HGCal Geometry

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

A step toward the V17 version of HGCal Geometry. The tools to get cell u,v from the position or the position from u,v are now part of the main link to SIM/RECO geometry - HGCalDDDConst. These codes take care of all types of cells (full and partial wafers) in the correct way. In earlier implementations of V16/V15/... the extended cells were not fully treated and parts of the cells could not be identified (x,y to u,v were not proper). The current implementation will take care of this - so the results of those versions will be marginally affected.

#### PR validation:

Tested using a script inside Geometry/HGCalCommonData/test/python

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by civanch** [comment]
@bsunanda , previous Phase-2 WFs D49, D60, D77 are also affected, results are different. Is this expected?

**Comment by bsunanda** [comment]
There was a bug in getting detid from position affecting all extended cells. This is taken care of here. Now the question is should one use the buggy software or ignore the difference

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 1]
```diff
@@ -183,15 +189,29 @@ std::array<int, 3> HGCalDDDConstants::assignCellTrap(float x, float y, float z,
   if (indx.first < 0)
     return std::array<int, 3>{{irad, iphi, type}};
   double xx = (z > 0) ? x : -x;
-  double r = (reco ? std::sqrt(x * x + y * y) : HGCalParameters::k_ScaleFromDDD * std::sqrt(x * x + y * y));
-  double phi = (r == 0. ? 0. : std::atan2(y, xx));
+  double phi = (((y == 0.0) && (x == 0.0)) ? 0. : std::atan2(y, xx));
   if (phi < 0)
     phi += (2.0 * M_PI);
+  iphi = 1 + (int)(phi / indx.second);
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:195]: `indx.second` is explicitly allowed to be 0. There should be protection against division by 0.
```suggestion
if (indx.second != 0)
  iphi = 1 + (int)(phi / indx.second);
```

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 2]
```diff
@@ -1519,55 +1581,63 @@ int HGCalDDDConstants::cellHex(
...
+    cellV = uv.second;
+  } else if (waferHexagon8File()) {
+    auto uv = hgcellUV_->cellUVFromXY3(xloc, yloc, place, cellType, extend, debug);
+    cellU = uv.first;
+    cellV = uv.second;
+  } else {
+    int N = (cellType == 0) ? hgpar_->nCellsFine_ : hgpar_->nCellsCoarse_;
+    double Rc = 2 * rmax_ / (3 * N);
+    double rc = 0.5 * Rc * sqrt3_;
+    double RcT = (extend) ? (2 * rmaxT_ / (3 * N)) : Rc;
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:1599]: Variable names should start with lowercase.

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:1599]: Trying. to take care of this

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:195]: Trying to take care of this

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
+1

**Comment by srimanob** [comment]
Hi @bsunanda 
Could you please update the PR description, mentioned about the bug and fix from this PR? Thanks.

**Comment by bsunanda** [comment]
@srimanob Please approve this PR

**Comment by bsunanda** [comment]
@perrotta @qliphy @srimanob Could you merge this PR urgently

**Comment by srimanob** [comment]
+Upgrade

This PR can be considered as an improvement on getting the cell position. It affects previous geometries also. This should be monitored when validation comes. FYI @cms-sw/pdmv-l2

**Comment by qliphy** [comment]
+1
a bug is fixed in getting detid from position affecting all extended cells,  Phase-2 WFs results get changed.

**Pull request merged**